jumanji/environments/logic/sliding_tile_puzzle/env.py

```diff
@@ -0,0 +1,266 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Sequence, Tuple, cast
+
+import chex
+import jax
+import jax.numpy as jnp
+import matplotlib.animation as animation
+from jax import lax
+from numpy.typing import NDArray
+
+from jumanji import specs
+from jumanji.env import Environment
+from jumanji.environments.logic.sliding_tile_puzzle.generator import (
+    Generator,
+    RandomGenerator,
+)
+from jumanji.environments.logic.sliding_tile_puzzle.reward import (
+    ImprovedDenseRewardFn,
+    RewardFn,
+)
+from jumanji.environments.logic.sliding_tile_puzzle.types import Observation, State
+from jumanji.environments.logic.sliding_tile_puzzle.viewer import (
+    SlidingTilePuzzleViewer,
+)
+from jumanji.types import TimeStep, restart, termination, transition
+from jumanji.viewer import Viewer
+
+
+class SlidingTilePuzzle(Environment[State]):
+    """Environment for the Sliding Tile Puzzle problem.
+
+    The problem is a combinatorial optimization task where the goal is
+    to move the empty tile around in order to arrange all the tiles in order.
+
+    - observation: `Observation`
+        - puzzle: jax array (int32) of shape (N, N), representing the current state of the puzzle.
+        - empty_tile_position: Tuple of int32, representing the position of the empty tile.
+        - action_mask: jax array (bool) of shape (4,), indicating which actions are valid
+            in the current state of the environment.
+
+    - action: Tuple of int32, representing the direction to move the empty tile
+        (up, down, left, right)
```

Comment:
Action is just an int

```suggestion
    - action: int32, representing the direction to move the empty tile
        (up, down, left, right)
```